### PR TITLE
Support CNI_ARGS in static IPAM plugin

### PR DESCRIPTION
**Pull request by s1061123**
This change is to add CNI_ARGS support in static IPAM plugin.
When IP/SUBNET/GATEWAY are given in CNI_ARGS, static IPAM adds
these info in addition to config files.

To configure ip address only from CNI_ARGS, 'address' field in config
is changed to optional from required.

### CODE REVIEW

File: plugins/ipam/static/README.md [hunk 1]
```diff
@@ -38,8 +38,17 @@ static IPAM is very simple IPAM plugin that assigns IPv4 and IPv6 addresses stat
...
 * `type` (string, required): "static"
-* `addresses` (array, required): an array of arrays of ip address objects:
+* `addresses` (array, optional): an array of arrays of ip address objects:
 	* `address` (string, required): CIDR notation IP address.
 	* `gateway` (string, optional): IP inside of "subnet" to designate as the gateway.
 * `routes` (string, optional): list of routes add to the container namespace. Each route is a dictionary with "dst" and optional "gw" fields. If "gw" is omitted, value of "gateway" will be used.
 * `dns` (string, optional): the dictionary with "nameservers", "domain" and "search".
+
+## Supported arguments
+The following [CNI_ARGS](https://github.com/containernetworking/cni/blob/master/SPEC.md#parameters) are supported:
```
Comment (jellonek) [inline plugins/ipam/static/README.md]: Missing new line before this one. Please keep that in sync with rest of document.

File: plugins/ipam/static/README.md [hunk 2]
```diff
@@ -38,8 +38,17 @@ static IPAM is very simple IPAM plugin that assigns IPv4 and IPv6 addresses stat
...
+
+* `IP`: request a specific IP address from a subnet
```
Comment (jellonek) [inline plugins/ipam/static/README.md]: Please remove ` from a subnet`.

File: plugins/ipam/static/README.md [hunk 3]
```diff
@@ -38,8 +38,17 @@ static IPAM is very simple IPAM plugin that assigns IPv4 and IPv6 addresses stat
 ## Network configuration reference
 
 * `type` (string, required): "static"
-* `addresses` (array, required): an array of arrays of ip address objects:
+* `addresses` (array, optional): an array of arrays of ip address objects:
```
Comment (jellonek) [inline plugins/ipam/static/README.md]: That's incorrect even in master - it's an array of ip addresses, not an array of arrays.

File: plugins/ipam/static/README.md (hunk 1 above)
Comment (s1061123) [reply plugins/ipam/static/README.md]: Fixed.

File: plugins/ipam/static/README.md (hunk 2 above)
Comment (s1061123) [reply plugins/ipam/static/README.md]: Fixed.

File: plugins/ipam/static/README.md (hunk 3 above)
Comment (s1061123) [reply plugins/ipam/static/README.md]: Fixed.

**Comment by squeed** [comment]
Seems fine, except it adds a new runtime convention. others might care about this.

**Comment by s1061123** [comment]
@squeed thank you for your comments. Let me clarify that your comment indicate to remove runtime convention, correct? I will remove runtime and update PR.

**Comment by s1061123** [comment]
@squeed I've taken care of your comments. Could you please take a look into it?

**Comment by dcbw** [comment]
@s1061123 thanks for the update, discussed this one during maintainer's meeting.  Instead of having two variables for IP and SUBNET, could we instead combine them into just IP that would have a CIDR format eg "1.2.3.4/24"?  You can give that to net.ParseCIDR() and you'll get both the IP address itself and the subnet mask out.

**Comment by s1061123** [comment]
@dcbw and other CNI committer, I need to modify a bit to support multiple ip address ("10.2.2.42/24" and "2001:db8::5/64").

**Comment by s1061123** [comment]
@dcbw , I updated the PR.

File: plugins/ipam/static/main.go [hunk 4]
```diff
@@ -109,6 +117,50 @@ func LoadIPAMConfig(bytes []byte, envArgs string) (*IPAMConfig, string, error) {
...
+		if e.GATEWAY != "" {
+			for _, item := range strings.Split(string(e.GATEWAY), ",") {
+				gwip := net.ParseIP(strings.TrimSpace(item))
+				if gwip == nil {
+					return nil, "", fmt.Errorf("invalid gateway address: %s", item)
+				}
+
+				for i := range n.IPAM.Addresses {
+					if n.IPAM.Addresses[i].Address.Contains(gwip) {
+						n.IPAM.Addresses[i].Gateway = gwip
```
Comment (dcbw) [inline plugins/ipam/static/main.go]: I would parse GATEWAY before IP, and hten assign the gateways to each Address while parsing the IPs.  We should also make sure that the GATEWAY array is the same length as the IP array.

File: plugins/ipam/static/main.go [hunk 5]
```diff
@@ -109,6 +117,50 @@ func LoadIPAMConfig(bytes []byte, envArgs string) (*IPAMConfig, string, error) {
...
+			return nil, "", err
+		}
+
+		if e.IP != "" {
+			for _, item := range strings.Split(string(e.IP), ",") {
+				ipstr := strings.TrimSpace(item)
+
+				ip, subnet, err := net.ParseCIDR(ipstr)
+				if err != nil {
+					return nil, "", fmt.Errorf("invalid CIDR %s: %s", e.IP, err)
```
[the remark this answers is not in the document]
Comment (s1061123) [reply plugins/ipam/static/main.go]: Sure. I will change it.

File: plugins/ipam/static/main.go (hunk 4 above)
Comment (s1061123) [reply plugins/ipam/static/main.go]: In some case, there is no gateway corresponding to certain IP address, especially multiple interface case. So I'd like to parse IP first then match gateway if there exists. Is that okey to you?

Comment (dcbw) [reply plugins/ipam/static/main.go]: @s1061123 yeah, you're right.

**Pull request merged**